### ISSUE TITLE
[HOLD for payment 2023-10-20] [$250] OldDot Transition - Opening free policy from OD with another user logged to ND opens blank page

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#26530]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Log in NewDot with an account
2. Navigate to OldDot and sign in to a different account that has a existing workspace.
3. Go to Settings > Policies > Group and click on any Workspace policy

## Expected Result:
You're redirected to NewDot, you're logged in with the same account as in OldDot and that the Workspace settings page is displayed

## Actual Result:
You're redirected to NewDot, you're logged in with the same account as in OldDot but the page is shown blank. User needs to refresh or use the link again for the page to appear correctly

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.61-3

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/5df9281e-4133-4a6f-be37-5a95c13f4654

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017c59d6ab1d74e2a2</li>
        <li>Upwork Job ID: 1698462755419181056</li>
        <li>Last Price Increase: 2023-10-16</li>
<li>Automatic offers: </li>
<ul>
<li>sourcecodedeveloper | Contributor | 27244329</li>
</ul></li>
    </ul>
</details>

**Comment by jliexpensify** [comment]
@lanitochka17 can you try this on v1.3.62-1?

I can't exactly reproduce this - basically clicking on a Workspace in Old.Dot opens up New.Dot, but there's considerable lag.

**Comment by lanitochka17** [comment]
Issue reproducible on Build 1.3.62-3

https://github.com/Expensify/App/assets/78819774/4005184c-e703-47fb-b344-0647d94e7ac3

**Comment by jliexpensify** [comment]
Ok yeah, I am seeing it now - thanks! Assigning to External.

**Comment by sourcecodedeveloper** [comment]
@lanitochka17 can you please verify the steps [here](https://github.com/Expensify/App/issues/26530#issuecomment-1706687452) and confirm the issue is still reproducible. 
@jliexpensify please advise how to test olddot code as it is not accessible to the public https://github.com/Expensify/App/issues/22769#issuecomment-1641462271

**Comment by jliexpensify** [comment]
Hmm the source code for www.expensify.com is not made publicly available. If this is a change that needs to occur on Old.Dot, then we would likely make this one `Internal` - I'll let @burczu weigh in when they are back.

**Comment by burczu** [comment]
@jliexpensify I think we should make it `Internal` - even if there is no need to change anything in OldDot eventually, I think it would be good to have a chance to check it.

**Comment by jliexpensify** [comment]
Hi @danieldoglas - this one looks like it needs to be handled Internally.

**Comment by danieldoglas** [comment]
I don't agree we need to fix this one internally though. 

The way we'll redirect from old dot to new dot will still be the same. We need new dot to clear up it's storage if it already is logged in when we are navigating from another account. This can be external @jliexpensify

**Comment by jliexpensify** [comment]
Still awaiting proposals!

**Comment by jliexpensify** [comment]
There are still no proposals, going to bump this one up to $1000 since it's been 3 weeks!

**Comment by jliexpensify** [comment]
Going top bump this up to $1000, since it's 3 weeks old. I've just re-tested this with 2 accounts and I'm seeing this:

![image](https://github.com/Expensify/App/assets/30815269/efe1f92d-a57f-431f-915a-96e5e3295cfc)

It's not a blank page, but the workspace is still not loading properly.

**Comment by b4s36t4** [comment]
@jliexpensify I'm a bit confused on expected behaviour here. When I'm coming from OD to ND the User B should automatically should login to ND? or we have to show login screen to User B (coming from OD)

As I see the in code we conditionally logout the user in which user should see the LoginScreen (which is not working at this point). Can you please explain the expected behaviour here?

**Comment by jliexpensify** [comment]
Hi @b4s36t4 how I understand this issue (and if I'm wrong, please correct me!):

- Log into New Expensify as user A
- Log into Expensify as user B
- From user B's account, click on a Workspace in the Policies page. This should open up New Expensify for User B's account and bring them to their Workspace

At the moment, it's opening up User B's account, but we see continued spinners or a blank page.

**Comment by jliexpensify** [comment]
Not overdue, waiting on proposals

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
1) Log in New Expensify using account A
2) Log in Expensify using account B
3) From account B in Expensify, clicking on workspace in Policies should open New Expensify for acccount B and open the workspace, but instead blank screen is shown

### What is the root cause of that problem?
The root cause is that storage is not being cleared in New Expensify when account A is already logged in and when the user is trying to navigate from account B logged in Expensify.

### What changes do you think we should make in order to solve the problem?
The function setUpPoliciesAndNavigate() in App.js needs to clear onyx storage by calling Onyx.clear()


```
// Sign out the current user if we're transitioning with a different user
const isTransitioning = Str.startsWith(url.pathname, Str.normalizeUrl(ROUTES.TRANSITION_BETWEEN_APPS));
if (isLoggingInAsNewUser && isTransitioning) {
    Session.signOut();
    Onyx.clear();
}
```


### What alternative solutions did you explore? (Optional)

**Comment by sourcecodedeveloper** [comment]
@jliexpensify @danieldoglas [proposal ](https://github.com/Expensify/App/issues/26530#issuecomment-1736386621) attached.

**Comment by jliexpensify** [comment]
@sourcecodedeveloper no need to tag us. @burczu will be reviewing proposals (as he is C+)

**Comment by danieldoglas** [comment]
@burczu can you please check the proposals?

**Comment by burczu** [comment]
@danieldoglas Yes, I'll review it today.

**Comment by burczu** [comment]
@sourcecodedeveloper In general your solution is simple and I guess it may fix the issue, but I'm not sure how can I test it with localhost - once clicking on the link in old dot, it obviously redirects to new dot not localhost... any advice on that?

**Comment by sourcecodedeveloper** [comment]
@burczu For testing on localhost I had to do some setup first.

Click on the link in old dot and then let the link open, then open console and look for api calls for OpenOldDotLink, screenshot attached

![screenshot_authtoken](https://github.com/Expensify/App/assets/141340971/33a75e30-f244-457e-8fe6-16839c7de2ea)

basically copy that authToken (shown in screenshot) 

Then for reference open
https://github.com/Expensify/App/blob/05aa0c317e12a230977188dd5afb18cf4015bcb1/src/pages/LogInWithShortLivedAuthTokenPage.js#L58

Comment out line 58 on your localhost and paste 
```
// Session.signInWithShortLivedAuthToken(email, shortLivedAuthToken);
const authToken=""; // paste that authToken shown in screenshot in the const authToken
Session.signInWithShortLivedAuthToken(email, authToken);            
NetworkStore.setAuthToken(authToken);
```

Also import NetworkStore 
`import * as NetworkStore from '../libs/Network/NetworkStore';
`

Then do the proposal part to clear the storage, that is go to 
https://github.com/Expensify/App/blob/cfad43ff96f2a104c51fd475e07f0980c4379fcd/src/libs/actions/App.js#L376C8-L376C27
the if block changes to 

```
if (isLoggingInAsNewUser && isTransitioning) {
    Session.signOut();
    Onyx.clear();
}
```

Make sure you get the authToken and place it in the const authToken. Back to redirect link in old dot, make sure localhost is logged in using another account than the old dot, click the link in old dot then press escape so the loading stops and then you can prepare the link with localhost. Then paste the redirect link with localhost.

If you did the setup correctly on localhost, then after the loading finishes the workspace should be loaded 

screenshot attached for reference
![screenshot_workspace](https://github.com/Expensify/App/assets/141340971/1b88296d-0cd3-4f0a-b5b1-4267e045d924)

**Comment by burczu** [comment]
@sourcecodedeveloper Thanks for detailed explanation - thanks to it I've managed to reproduce the issue locally and test your proposed solution - it works indeed. 

That being said I think we could proceed with [the proposal](https://github.com/Expensify/App/issues/26530#issuecomment-1736386621) from @sourcecodedeveloper here.

🎀 👀 🎀 C+ reviewed

**Comment by sourcecodedeveloper** [comment]
> @sourcecodedeveloper Thanks for detailed explanation - thanks to it I've managed to reproduce the issue locally and test your proposed solution - it works indeed.
> 
> That being said I think we could proceed with [the proposal](https://github.com/Expensify/App/issues/26530#issuecomment-1736386621) from @sourcecodedeveloper here.
> 
> 🎀 👀 🎀 C+ reviewed

Thanks for accepting the proposal. I will start preparing the PR and get back shortly.

**Comment by burczu** [comment]
@sourcecodedeveloper I think we need to wait for @danieldoglas final approval first.

**Comment by sourcecodedeveloper** [comment]
> @sourcecodedeveloper I think we need to wait for @danieldoglas final approval first.
Great. Once approved, I can start preparing the PR.

**Comment by danieldoglas** [comment]
LGTM

**Comment by sourcecodedeveloper** [comment]
Great I shall have the PR ready shortly.

**Comment by sourcecodedeveloper** [comment]
@burczu PR is ready for review https://github.com/Expensify/App/pull/28819

**Comment by koko57** [comment]
looks like a very similar issue to this one https://github.com/Expensify/App/issues/27418 
fixed in: https://github.com/Expensify/App/pull/28984

**Comment by burczu** [comment]
@koko57 Thanks for pointing this out - I'll check if it's fixed by the PR you've mention soon.

**Comment by sourcecodedeveloper** [comment]
@burczu I just want to mention that I was first to propose the solution. The proposal in PR (https://github.com/Expensify/App/pull/28984) was made on Oct 4th and I made the proposal on Oct 2nd.

**Comment by sourcecodedeveloper** [comment]
@burczu I have taken quick look at PR (https://github.com/Expensify/App/pull/28984) and it has made some breaking changes. It also breaks deeplinking for the web. The transition link is supposed to open in app when clicked from the chat window but instead I get loading spinner, screenshot attached.

![screenshot_spinner](https://github.com/Expensify/App/assets/141340971/6b5881b0-b6e6-48a5-ba48-97350f809829)

LogOutPreviousUserPage component is no longer logging out the user but to the reader it appears that the component is intended to handle signout. If the component is no longer needed it could be removed.

The PR(https://github.com/Expensify/App/pull/28984) has the possibility to cause regression. Deeplinking for android/web and ios/web would also require testing.

I suggest that the PR(https://github.com/Expensify/App/pull/28819) be merged first because the code changes are simple to understand and do not break deeplinking for web, android-web or ios-web.

Once PR(https://github.com/Expensify/App/pull/28819) is merged, the other PR(https://github.com/Expensify/App/pull/28984) can be reviewed and merged.

**Comment by sourcecodedeveloper** [comment]
@koko57 The deep linking breaks as shown by the loading spinner in [screenshot](https://github.com/Expensify/App/pull/28984). The transition link is supposed to open in chat window and in your PR Author checklist you would need to post videos for remaining platforms. 

@burczu Doing that would require considerable time. What I suggest is PR https://github.com/Expensify/App/pull/28819 be merged first and in the meantime @koko57 can take the time and test deep linking on the remaining platforms and then merge their PR.

**Comment by koko57** [comment]
@sourcecodedeveloper how did you get the spinner? What are the exact step to reproduce? As I've mentioned I tried to reproduce this flow and it worked fine.

**Comment by sourcecodedeveloper** [comment]
@burczu Can you please provide some feedback on the discussion and how to proceed.

**Comment by burczu** [comment]
@sourcecodedeveloper Sorry, I had no time yet to come back to this issue and take a closer look on both solutions... I'll get back to it once I finish my other tasks. 

In terms of reproducing this issue, I've followed your advices from this comment: https://github.com/Expensify/App/issues/26530#issuecomment-1743466000

**Comment by burczu** [comment]
@koko57 @sourcecodedeveloper I've checked the solution provided in https://github.com/Expensify/App/pull/28984 and it looks like it fixes this issue. In my opinion this other solution looks better because the check is done in `AuthScreens` so it's done earlier. Additionally it seem to work a little bit faster, so I think we should go with that other solution and close this one.

@danieldoglas @jliexpensify I case we agreed with the above, I think we should compensate @sourcecodedeveloper for his effort here, what do you think?

**Comment by sourcecodedeveloper** [comment]
@burczu the solution in the other PR breaks deeplinking I suggest, the current PR be tested and merged and let the other PR resolve deeplinking at least then later merge after the current PR is merged.

**Comment by burczu** [comment]
@sourcecodedeveloper In terms of breaking deeplinking - are you referring to this statement of yours?

> The transition link is supposed to open in app when clicked from the chat window but instead I get loading spinner, screenshot attached.

I've just tried it and it works for me.

**Comment by sourcecodedeveloper** [comment]
@burczu If you follow the flow shown in the video, you should be able to replicate the crash. 

https://github.com/Expensify/App/assets/141340971/f2e82fdf-b559-422a-9f42-9c57faf272c3

**Comment by burczu** [comment]
@sourcecodedeveloper I see your point but I'm not sure this is the real life scenario we need to handle in the app (the `/transition` route isn't supposed to be used in other way than while switching from OldDot to NewDot). I've [posted a comment](https://github.com/Expensify/App/pull/28984#issuecomment-1764224148) in the other PR asking others about their opinion.

**Comment by sourcecodedeveloper** [comment]
@burczu how else would the transition link be tested on android / ios -web and also native. When preparing videos for ios and android -web, the approach i used for testing transition link was using deeplinking.

**Comment by sourcecodedeveloper** [comment]
@burczu I was reading over some previous comments, besides loading the workspace, the spinners and or blank page were also part of the issue as mentioned in the [comment](https://github.com/Expensify/App/issues/26530#issuecomment-1731190641). 
If the other PR is merged, the issue of spinner would still be there, that is why I suggest the current PR be merged first and then the other can be merged.

**Comment by mountiny** [comment]
@jliexpensify as a payment summary we got $250 to @sourcecodedeveloper for their work on this issue, we are going to use different solution from a different issue which has similar effects though.

**Comment by sourcecodedeveloper** [comment]
@mountiny Thanks for stepping in. The 25% rule happened in another PR as well, where the PR fix was made and then half way during review the some other author merged directly in main branch and the reviewer was unable to test the PR. As per contributor guidelines, I wanted to request re-evaluation of the 25% compensation because it is the second time similar has happened and such practice should be discouraged given that I had the proposal awarded first and the other solution could have been discussed prior to the PR being created.

**Comment by mountiny** [comment]
> I wanted to request re-evaluation of the 25% compensation because it is the second time similar has happened and such practice should be discouraged given that I had the proposal awarded first and the other solution could have been discussed prior to the PR being created.

I understand this is not ideal situation and I am sorry for this. In ideal world there would be no duplicates, however with the number of issues and contributors it happens that we miss this and there might be similar issue/ solution happening else where. 

I dont want to make exceptions here as that would set precedence for other issues and makes the rules muddier, will stick with 25% here and I know you are a great engineer and contributor, who will be able to fix many other issues which will compensate for these two issues. This does not happen often and I am sorry you had this happen to you twice.

**Comment by sourcecodedeveloper** [comment]
@mountiny I agree that the situation is not ideal but given that the proposal was selected first, the later PR could have either been merged with the existing PR or could have waited, it does not even solve the problem of the loading spinner as indicated in the issue earlier. I would still like to get the opinion be heard by someone else is there anybody else who can decide on the 25% besides yourself, please advise.

**Comment by sourcecodedeveloper** [comment]
@mountiny the issue of loading spinner has been [raised](https://github.com/Expensify/App/pull/28984#issuecomment-1764774500) and the expectation was the other pr would resolve the loading spinner. Please re consider merging the PR first as it would resolve the spinner and the workspace issue. or if there is anything else you would want changed in the PR, please discuss.

**Comment by mountiny** [comment]
I believe @koko57 has found a fix for the spinner now https://github.com/Expensify/App/pull/28984

**Comment by jliexpensify** [comment]
Based on this [comment](https://github.com/Expensify/App/issues/26530#issuecomment-1764376250), is this correct for a final Payment Summary for this issue @mountiny ?

- Contributor: @sourcecodedeveloper $250 (can you edit your GH profile to include the same name as your Upworks profile, and apply for https://www.upwork.com/jobs/~017c59d6ab1d74e2a2?)

I'm assuming that we can sloe this issue after payment in favour of https://github.com/Expensify/App/pull/28984?

**Comment by sourcecodedeveloper** [comment]
@mountiny can you please try and trigger automatic hiring, really the automation should have sent an offer [link](https://github.com/Expensify/App/issues/27085#issuecomment-1723179854)

**Comment by sourcecodedeveloper** [comment]
@mountiny Thanks.

**Comment by jliexpensify** [comment]
Paid and job closed!
